[8.x] Add anonymous migrations

thiagorb opened the pull request:
This would solve https://github.com/laravel/framework/issues/5899.

With this approach, migrations won't have class names anymore (they didn't serve any purpose and cause some tricky to solve issues).

After a project was maintained for a long time, the chance that two migrations were created with the same name is very high. For example, consider the following migrations: 
- 2019_01_01_000000_add_content_to_posts_table
- 2020_01_01_000000_remove_content_from_posts_table
- 2021_01_01_000000_add_content_to_posts_table

With the current approach, the first and last migration have the same name. After the third migration is created, it is possible to run it just fine, unless you are trying to run all these migrations at the same time (like after creating an empty database, or while running tests). In a project where there were always tests since the beginning this will be barely an inconvenience, but it causes a lot of trouble for a project that has been developed for many years without tests.

With this new approach this won't be a problem, cause the names can be dropped completely. The new Migrator is still able to run named migration classes, so the solution is backward compatible.

I believe it would be possible to backport it to 6.x too, in case it is approved.

thiagorb committed:
Extend MigratorTest

thiagorb committed:
Implement anonymous migrations

crynobone commented:
This PR changes some of the framework migrations such as `batches`, `failed_jobs`. Is there anyway preventing the app from generating the migration more than once?

![image](https://user-images.githubusercontent.com/172966/113962943-4743be80-985b-11eb-952c-3e9824111503.png)

Also wouldn't think we should change the default stubs on a patch release, but supporting anonymous migrations itself on patch release should be fine if there no breaking change risk 👍🏼 .

thiagorb commented:
Hey @crynobone ! Yeah, I thought that changing the stubs could be too much, so I added it as a separate commit. We can drop this last commit then.

But your preference is to throw if creating a migration with duplicated name instead of anonymous migrations?

crynobone commented:
> But your preference is to throw if creating a migration with duplicated name instead of anonymous migrations?

I worry that it potentially causes an issue if developers try to run `php artisan queue:failed-table` more than once since it should generate the same file with a different timestamp and no longer check for the identical class name as if right now (see above picture).

My guess all framework migration should remain as it is. and maybe ship with separate anonymous migration stubs for 8.x. 

```bash
php artisan make:migration -a create_posts_table --create
```

driesvints commented:
I really like this idea. I think @crynobone indeed makes a good point about the duplicate framework shipped migrations. 

> I believe it would be possible to backport it to 6.x too, in case it is approved.

Laravel 6 is closed for new features.

taylorotwell commented:
It's a cool feature but yeah I would drop the commit of renaming existing migrations.

thiagorb commented:
Ok. I reverted the commit with the stubs update.

taylorotwell commented:
There is a breaking change by changing the public `resolve` method on migrator to `protected`.

driesvints commented on src/Illuminate/Database/Migrations/Migrator.php:
```diff
@@ -441,16 +435,23 @@ protected function getQueries($migration, $method)
      * @return object
      */
-    public function resolve($file)
+    protected function resolve(string $path)
```
We can't update this to being a path. Apps will break if they expect to pass a file instead. All these changes to this method need to be reverted and a different solution needs to be implemented.

thiagorb replied:
I changed the code to have the new implementation in the `resolvePath` method. This method `resolve` was adjusted so that it is public again, and keeps the old behavior, even though it is not used in the Migrator anymore. Still, I changed the name of the parameter, because what we give to this method is actually a migration name (like "2020_01_01_000000_create_users_table"), not a file name.

I also restored the function `requireFiles`. It is also not used anymore, but in theory it is a breaking change.

thiagorb commented:
> There is a breaking change by changing the public `resolve` method on migrator to `protected`.

I restored the public method, and created a new protected one for the new implementation. The old method is not being used anymore though.

driesvints commented on src/Illuminate/Database/Migrations/Migrator.php:
```diff
@@ -404,16 +400,14 @@ protected function runMigration($migration, $method)
      * @return void
      */
-    protected function pretendToRun($migration, $method)
+    protected function pretendToRun($migrationName, $queries): void
```
Please remove all return types. Those are breaking changes to a method signature.

driesvints replied:
Also, the behavior of this method changed. You can't change the parameter types or what this method does. Please revert all changes to existing methods so they behave like before.

thiagorb replied:
This method's behavior was restored.

rodrigopedra commented on src/Illuminate/Database/Migrations/Migrator.php:
```diff
@@ -441,18 +435,49 @@ protected function getQueries($migration, $method)
      * @return object
      */
-    public function resolve($file)
+    public function resolve($migrationName)
```
As Laravel 8.x supports PHP 8, changing a public method parameter name is a breaking change, due to named parameters feature introduced in PHP 9

driesvints replied:
@rodrigopedra changing parameter names is an exception to our BC policy: https://laravel.com/docs/8.x/releases#named-arguments

driesvints replied:
That being said, yes please revert this.

rodrigopedra replied:
Thanks @driesvints , I was not aware of that

thiagorb replied:
@driesvints I'm confused, should I keep `$migrationName` or revert to `$file`. As I mentioned before, what is being passed to this method is not a file, but a migration name.

driesvints replied:
@thiagorb let's please focus on the feature itself without too much distraction.

thiagorb replied:
The paramter's name was reverted.

thiagorb committed:
Display migration class name in pretendToRun

Due to backward compatibility, classic migrations (non-anonymous) should
have their class name displayed while running migrations with --pretend,
instead of the migration name, like anywhere else in the code.

thiagorb commented:
I just pushed one more change, cause I realized that while running `php artisan migrate --pretend`, the queries would be printed differently:
```
2014_10_12_000000_create_people_table: create table "people"...
```
Instead of:
```
CreatePeopleTable: create table "people"...
```
For the anonymous migrations I left it as:
```
2015_10_04_000000_modify_people_table: alter table "people" add column
```
It would be easy to transform the name into `ModifyPeopleTable` (by reusing existing methods), but I think it doesn't make sense to create this class name if the class doesn't exist. Besides, the class name is displayed only while running with `--pretend`, so I think it would be better to change it in the future to always display the migration name, to be consistent with how it is displayed while migrating or rolling back for real:
```
Migrating: 2014_10_12_000000_create_people_table
```
```
Rolling back: 2014_10_12_000000_create_people_table
```

driesvints commented:
@thiagorb this is a fantastic PR. Thanks for your work 👍

lloricode commented:
this should be mandatory in laravel 9

lloricode commented:
> this should be mandatory in laravel 9
> 
> edit:
> so migrations in laravel 9 won't have class names anymore

hope to merge this #37352

BenjaminStuermer commented:
Hey, @thiagorb, I just came across this PR after running into exactly the issue you describe: I developed a project migration by migration, had multiple identical migration names, and never noticed until I tried to create a new instance of the application. You saved me some serious pain with this PR, thank you so much.
